### PR TITLE
Added support for toggling enter key requirement

### LINKED ISSUE #83: Do not require enter on yes/no questions
It's a little detail, but it's something pacaur does I quite like. When asking questions about pkgbuilds, etc, pacaur only requires you to press y/n and moves on immediately. Pikaur currently requires you to type your input and then press the enter key, which just feels slower than it should be.

### PR DESCRIPTION
**Pull request by JakeStanger**
Currently a constant at the top of `prompt`, a boolean can be toggled to switch between built-in `input` function for requiring enter key or custom `get_answer` function which does not require enter key confirmation.

Requires integrating with the config file.

_edited_:
closes #83

### CODE REVIEW

**Comment by actionless** [comment]
you can see here how to add a new config option:
https://github.com/actionless/pikaur/commit/fb77af54e5ff1edf488628a14fdc6f0b3fb65eed

also i see travis is failing for the PR

**Comment by JakeStanger** [comment]
Probably would have been clever to have waited until the morning and checking I'd done a proper job rather than submitting a PR at 3am. Hold tight.

File: pikaur/config.py [hunk 1]
```diff
@@ -56,6 +56,12 @@
             'type': 'int',
             'default': '9',
         },
+    },
+    'misc': {
```
Comment (actionless) [inline pikaur/config.py]: mb `ui` instead of `misc`?

File: pikaur/prompt.py [hunk 2]
```diff
@@ -1,26 +1,76 @@
 import sys
+import tty
 from typing import Callable, List
 
+from pikaur.config import PikaurConfig
+
+from pikaur.args import PikaurArgs
```
Comment (actionless) [inline pikaur/prompt.py]: i do prefer relative imports

File: pikaur/config.py (hunk 1 above)
Comment (JakeStanger) [reply pikaur/config.py]: That's probably a more reasonable name, yeah. I blanked.

File: pikaur/prompt.py [hunk 3]
```diff
@@ -1,26 +1,76 @@
...
 from .core import interactive_spawn
 from .i18n import _
 from .pprint import color_line, print_status_message
-from .args import PikaurArgs
 
 
-def ask_to_continue(text: str = None, default_yes=True, args: PikaurArgs = None) -> bool:
-    if text is None:
-        text = _("Do you want to proceed?")
+def get_answer(question, answers='Yn'):
```
Comment (actionless) [inline pikaur/prompt.py]: please add typing hint for the first argument

File: pikaur/prompt.py (hunk 2 above)
Comment (JakeStanger) [reply pikaur/prompt.py]: Did not know that was a thing you could do to be honest, changed those.

File: pikaur/prompt.py (hunk 3 above)
Comment (JakeStanger) [reply pikaur/prompt.py]: Should `text` having a default in the parameters not cover that?

File: pikaur/prompt.py [hunk 4]
```diff
@@ -1,26 +1,75 @@
...
+        answer = input(prompt).lower()
+    else:
+        answer = get_answer(prompt, answers=answers)
+
+    return answer
+
+
+def ask_to_continue(text='Do you want to proceed?', default_yes=True, args: PikaurArgs=None) -> bool:
     if args and args.noconfirm and default_yes:
         print_status_message('{} {}'.format(text, _("[Y]es (--noconfirm)")))
```
Comment (actionless) [inline pikaur/prompt.py]: this part seems broken so it causing an exception during travis

Comment (actionless) [reply pikaur/prompt.py]: basically you just need to return the next line you removed

File: pikaur/prompt.py (hunk 3 above)
Comment (actionless) [reply pikaur/prompt.py]: `def get_answer(question, answers='Yn'):`

Comment (JakeStanger) [reply pikaur/prompt.py]: Ooh sorry I see what you mean now. Do you want all parameters hinted or just ones without defaults?

Comment (actionless) [reply pikaur/prompt.py]: also return values not hinted as well

File: pikaur/prompt.py [hunk 5]
```diff
@@ -1,26 +1,75 @@
...
+    except Exception:
+        return ' '
+    finally:
+        tty.tcsetattr(sys.stdin.fileno(), tty.TCSADRAIN, previous_tty_settings)
+        sys.stdout.write('{}\r\n'.format(answer))
+        tty.tcdrain(sys.stdin.fileno())
+
+
+def get_input(prompt: str, answers: str=None):
+    if PikaurConfig().misc.get('RequireEnterConfirm'):
```
Comment (actionless) [inline pikaur/prompt.py]: here you didn't updated `misc` -> `ui`

File: pikaur/install_cli.py [hunk 6]
```diff
@@ -381,12 +381,14 @@ def _print_sysupgrade(verbose=False) -> None:
 
         def _confirm_sysupgrade(verbose=False) -> str:
             _print_sysupgrade(verbose=verbose)
-            answer = input('{} {}\n{} {}\n> '.format(
+            prompt = '{} {}\n{} {}\n>>'.format(
                 color_line('::', 12),
-                bold_line(_("Proceed with installation? [Y/n]")),
+                bold_line('Proceed with installation? [Y/n] '),
```
[the remark this answers is not in the document]
Comment (JakeStanger) [reply pikaur/install_cli.py]: Not sure how I managed that one considering I copied and pasted it honestly. I'm making you do far more work than you should have to here, sorry.

File: pikaur/prompt.py [hunk 7]
```diff
@@ -61,7 +61,10 @@ def get_input(prompt: str, answers: str=None) -> str:
     return answer
 
 
-def ask_to_continue(text: str='Do you want to proceed?', default_yes: bool=True, args: PikaurArgs=None) -> bool:
+def ask_to_continue(text: str, default_yes: bool=True, args: PikaurArgs=None) -> bool:
+    if text is None:
+        text = 'Do you want to proceed?'
```
Comment (actionless) [inline pikaur/prompt.py]: should be _('string'), and also text is not defaulted as None

File: pikaur/prompt.py [hunk 8]
```diff
@@ -1,26 +1,78 @@
...
+
+
+def ask_to_continue(text: str, default_yes: bool=True, args: PikaurArgs=None) -> bool:
     if text is None:
-        text = _("Do you want to proceed?")
+        text = 'Do you want to proceed?'
+
     if args and args.noconfirm and default_yes:
         print_status_message('{} {}'.format(text, _("[Y]es (--noconfirm)")))
-        return True
```
Comment (actionless) [inline pikaur/prompt.py]: why this line was removed?

File: pikaur/prompt.py [hunk 9]
```diff
@@ -1,26 +1,78 @@
...
+    else:
+        answer = get_answer(prompt, answers=answers)
+
+    return answer
+
+
+def ask_to_continue(text: str, default_yes: bool=True, args: PikaurArgs=None) -> bool:
     if text is None:
-        text = _("Do you want to proceed?")
+        text = 'Do you want to proceed?'
```
Comment (actionless) [inline pikaur/prompt.py]: internatialization

File: pikaur/prompt.py (hunk 8 above)
Comment (JakeStanger) [reply pikaur/prompt.py]: Quite honestly just completely missed it.

File: pikaur/prompt.py (hunk 4 above)
Comment (JakeStanger) [reply pikaur/prompt.py]: Christ I just realised it's internationalization not internalization. Sorry.

File: pikaur/prompt.py [hunk 10]
```diff
@@ -1,26 +1,79 @@
...
+        answer = get_answer(prompt, answers=answers)
+
+    return answer
+
+
+def ask_to_continue(text: str=None, default_yes: bool=True, args: PikaurArgs=None) -> bool:
     if text is None:
-        text = _("Do you want to proceed?")
+        text = _('Do you want to proceed?')
+        return True
```
Comment (actionless) [inline pikaur/prompt.py]: you moved it instead of just restoring so the travis still failing

File: pikaur/prompt.py [hunk 11]
```diff
@@ -1,26 +1,79 @@
...
+
     if args and args.noconfirm and default_yes:
         print_status_message('{} {}'.format(text, _("[Y]es (--noconfirm)")))
-        return True
```
Comment (actionless) [inline pikaur/prompt.py]: here

Comment (JakeStanger) [reply pikaur/prompt.py]: I edited the previous commits and lost this one in the process, and then re-committed it in the wrong place.

**Comment by actionless** [comment]
good, so we have two things to be finished and good to go!
1) few pylint warnings in the new functions https://travis-ci.org/actionless/pikaur/builds/358106633#L2499
2) short option description in readme

**Comment by JakeStanger** [comment]
Looks like the warnings are a false positive. Perhaps where both a type and a default value are specified?

File: pikaur/prompt.py [hunk 12]
```diff
@@ -1,26 +1,79 @@
...
+from .config import PikaurConfig
+
 from .core import interactive_spawn
 from .i18n import _
 from .pprint import color_line, print_status_message
-from .args import PikaurArgs
 
 
-def ask_to_continue(text: str = None, default_yes=True, args: PikaurArgs = None) -> bool:
+def get_answer(question: str, answers: str='Yn') -> str:
```
Comment (actionless) [inline pikaur/prompt.py]: and also mb rename this to `get_answer_from_tty`/`read_answer_from_tty` or smth like `get_answer_interactively`

Comment (JakeStanger) [reply pikaur/prompt.py]: I've also tried removing the type hint on `answers` to see if the issue with the build bot is that.

**Comment by actionless** [comment]
ah, you need to put the spaces like `answers: str = 'Yn')` to make it happy

File: pikaur/prompt.py (hunk 12 above)
Comment (JakeStanger) [reply pikaur/prompt.py]: Error went away, so that's the problem. Do you want me to stick it back in, or remove the others so the build officially passes?

**Comment by JakeStanger** [comment]
Ah I see! That's not in PEP8 or anything so I don't know why pylint checks for that.

**Comment by actionless** [comment]
it's PEP484, search it be `:.*=.*`

**Comment by actionless** [comment]
aaalmost :cat2:

**Comment by JakeStanger** [comment]
Well today I learned, thanks.

**Comment by actionless** [comment]
seems like travis bugged, restarting the build

**Comment by JakeStanger** [comment]
It would appear Travis is just having an off day - no errors, no warnings, 10.0/10, but still failed.

**Comment by actionless** [comment]
ok, so we dealed with pylint, but now mypy hates you:

https://travis-ci.org/actionless/pikaur/builds/358112525#L2504

**Comment by JakeStanger** [comment]
Is it using a different package version or something? All of those are built-ins that definitely exist in 3.6.

**Comment by actionless** [comment]
it seems what typeshed stubs are just missing for `tty` :sweat_smile: 

i guess only `  # type: ignore` comments would help :crying_cat_face:

**Comment by actionless** [comment]
:tada: :champagne: :evergreen_tree: :smoking:

**Comment by actionless** [comment]
FFFUUU, i've pressed rebase instead of squash :man_facepalming: :woman_facepalming:

**Comment by actionless** [comment]
thanks for your effort on finishing this feature for the release level!

**Pull request merged**